Fix fast_gnp_random_graph for directed graphs (issue #3389)

jamestrimble opened the pull request:
This fixes #3389 with the change suggested by @proto-n in that issue.  I have generated hundreds of thousands of graphs to confirm that the edge counts are as expected.

rossbar commented:
It'd be great if we could add a test for this to accompany the fix. It looks like it should be possible to adopt the snippet for checking uniformity from the original issue into a test (perhaps by seeding `fast_gnp_random_graph` with the loop variable `i`). Maybe there's a more elegant way to test it - other ideas welcome!

jamestrimble commented:
@rossbar good point. I've added a test for all the G(n,p) generators to check that the edge probabilities are roughly correct.

I've also simplified the algorithm. Now, to generate a directed graph, it just uses the same algorithm as the undirected generator, and does the same thing once more with v and w reversed. This seems much simpler, and also closer to the algorithm in the paper.

rossbar commented on networkx/generators/tests/test_random_graphs.py:
```diff
@@ -247,6 +247,29 @@ def test_gnp(self):
                 edges += sum(1 for _ in generator(10, 0.99999, directed=True).edges())
             assert abs(edges / float(runs) - 90) <= runs * 2.0 / 100
 
+            # assert that edges are generated with correct probability
+            runs = 5000
+            n = 5
+            for p in [0.2, 0.8]:
+                for directed in [False, True]:
+                    edge_counts = [[0] * 5 for row in range(5)]
+                    for i in range(runs):
+                        G = nx.generators.random_graphs.fast_gnp_random_graph(
+                            n, p, directed=directed
+                        )
+                        for (v, w) in G.edges:
+                            edge_counts[v][w] += 1
+                            if not directed:
+                                edge_counts[w][v] += 1
+                    for v in range(n):
+                        for w in range(n):
+                            if v == w:
+                                # There should be no loops
+                                assert edge_counts[v][w] == 0
+                            else:
+                                # Each edge should have been generated with probability close to p
+                                assert abs(edge_counts[v][w] / float(runs) - p) <= 0.03
```
As currently defined, this will run 4 times since it is inside the for-loop over random graph generators. The easiest way to fix this would be to move it into it's own test function (or method in this case).

jamestrimble replied:
Thanks for spotting that. I had intended to run the test for all four generators, but using a specific generator on line 257 was definitely a mistake. I've fixed it now so it uses the loop variable `generator` as the function.

Hopefully this is okay? If you think it's a waste of time testing the simpler generators I can move it outside the loop.

rossbar reviewed (approved):
LGTM, I think the new test could be iterated on a bit to make it more readable and use more pytest features, but honestly I think the entire test module could use some work so I'm inclined to push that off into another issue to track separately so it's not a blocker here.

Thanks @jamestrimble !

dschult reviewed (approved):
Thanks for implementing this @jamestrimble !

The tests in this module look like they need a unified look-over. Many use seeds for the tests to avoid intermittent failures. But others don't use seeds. We should figure out when seeds should be used and when the tests should be random.  We'll do that in another PR.

This change looks good to me.
